crates/db_schema/src/impls/tag.rs

```diff
@@ -22,6 +24,75 @@ impl Tag {
       .await
       .with_lemmy_type(LemmyErrorType::NotFound)
   }
+
+  pub async fn community_override_all_from_apub(
+    pool: &mut DbPool<'_>,
+    community_id: CommunityId,
+    community_ap_id: String,
+    ap_tags: Vec<TagInsertForm>,
+  ) -> LemmyResult<()> {
+    // Verify that each tag is actually in the given community.
+    // Needed to ensure that incoming AP updates of one community can not manipulate tags in a
+    // different community.
+    let ap_tags: Vec<TagInsertForm> = ap_tags
+      .into_iter()
+      .filter(|tag| tag.ap_id.as_str().starts_with(&community_ap_id))
+      .collect();
+
+    let known_tags = Tag::get_by_community(pool, community_id).await?;
+    let old_tags = known_tags
+      .iter()
+      .map(|tag| (tag.ap_id.clone(), tag))
+      .collect::<HashMap<_, _>>();
+    let new_tag_ids = ap_tags
+      .iter()
+      .map(|tag| tag.ap_id.clone())
+      .collect::<HashSet<_>>();
+
+    let to_delete = known_tags
+      .iter()
+      .filter(|tag| !new_tag_ids.contains(&tag.ap_id))
+      .map(|tag| tag.id)
+      .collect::<Vec<_>>();
+    let to_insert = ap_tags
+      .iter()
+      .filter(|tag| !old_tags.contains_key(&tag.ap_id))
+      .collect::<Vec<_>>();
+    for tag in to_insert {
+      Tag::create(pool, tag).await?;
+    }
+    // if display name is different, we need to update it
+    for tag in ap_tags {
+      if let Some(old_tag) = old_tags.get(&tag.ap_id) {
+        if old_tag.display_name != tag.display_name {
+          Tag::update(
+            pool,
+            old_tag.id,
+            &TagUpdateForm {
+              display_name: Some(tag.display_name.clone()),
+              updated: Some(Some(Utc::now())),
+              ..Default::default()
+            },
+          )
+          .await?;
+        }
+      }
+    }
```

Comment:
- This shouldnt need any extra logic, just insert all tags which are not deleted. 
- Its inefficient to make db calls in a loop, instead pass a [vec of forms with on conflict](https://github.com/LemmyNet/lemmy/blob/main/crates/db_schema/src/impls/person_post_mention.rs#L30)
- Should be inside a transaction

Reply:
Not sure if transaction is necessary, since you can do multiple upserts in a single statement `.values(forms)`, but ya upsert should be done.